File: ProblemSet/70.climbing-stairs/java/70.climbing-stairs 记忆化迭代.java
/*
 * @lc app=leetcode id=70 lang=java
 *
 * [70] Climbing Stairs
 */

// @lc code=start
class Solution {
    public int climbStairs(int n) {
        if (n <= 2) return n;
        int pre1 = 1, pre2 = 2, cur = 3;
        for (int i = 3; i <= n; ++i) {
            cur = pre1 + pre2;
            pre1 = pre2;
            pre2 = cur;
        }
        return cur;
    }
}
// @lc code=end

File: ProblemSet/70.climbing-stairs/python/70.climbing-stairs 记忆化迭代.py
#
# @lc app=leetcode id=70 lang=python3
#
# [70] Climbing Stairs
#

# @lc code=start
class Solution:
    def climbStairs(self, n: int) -> int:
        if n <= 2:
            return n
        pre1, pre2 = 1, 2
        for _ in range(3, n + 1):
            pre1, pre2 = pre2, pre1 + pre2
        return pre2
        
# @lc code=end

